src/config/vars.rs

```diff
@@ -16,27 +21,59 @@ pub fn interpolate(input: &str, vars: &HashMap<String, String>) -> (String, Vec<
         r"(?x)
         \$\$|
         \$([[:word:].]+)|
-        \$\{([[:word:].]+)(?::-([^}]+)?)?\}",
+        \$\{([[:word:].]+)(?:(:?-|:?\?)([^}]*))?\}",
     )
     .unwrap();
+
     let interpolated = re
         .replace_all(input, |caps: &Captures<'_>| {
+            let flags = caps.get(3).map(|m| m.as_str()).unwrap_or("");
+            let def_or_err = caps.get(4).map(|m| m.as_str()).unwrap_or("");
             caps.get(1)
                 .or_else(|| caps.get(2))
                 .map(|m| m.as_str())
                 .map(|name| {
-                    vars.get(name).map(|val| val.as_str()).unwrap_or_else(|| {
-                        caps.get(3).map(|m| m.as_str()).unwrap_or_else(|| {
-                            warnings.push(format!("Unknown env var in config. name = {:?}", name));
-                            ""
+                    vars.get(name)
+                        .map(|val| match val.as_str() {
+                            "" => match flags {
+                                ":-" => def_or_err,
+                                ":?" => {
+                                    errors.push(format!(
+                                        "Non-empty env var required in config. name = {:?}, error = {:?}",
+                                        name, def_or_err
+                                    ));
+                                    ""
+                                }
+                                _ => "",
+                            },
+                            val => val,
+                        })
+                        .unwrap_or_else(|| match flags {
+                            ":-" | "-" => def_or_err,
+                            ":?" | "?" => {
+                                errors.push(format!(
+                                    "Unkown env var in config. name = {:?}, error = {:?}",
+                                    name, def_or_err
+                                ));
+                                ""
+                            }
+                            _ => {
+                                warnings
+                                    .push(format!("Unknown env var in config. name = {:?}", name));
+                                ""
+                            }
                         })
```

Comment:
Would this logic be more straightforward to match on the `flags` at the top level instead of mapping the value?

Reply:
For the record, I was also not 100% happy about this part of the code :)
I did think of matching on `flags` first, however I got a bit tangled with handling the `HashMap::get()` because of the potential item-not-found case. I will give it another thought now that I am getting more familiar with Rust.

Thanks for the feedback!

Reply:
Hi @bruceg, I managed to implement the code by matching on `flags` instead of mapping the var values as suggested.

The code could be cleaner with these alternatives that unfortunately are not ready in Rust:
* The ["if-let-chains" RFC](https://rust-lang.github.io/rfcs/2497-if-let-chains.html), which is still pending.
* The [`is_some_with()` method](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_with), which is still experimental (and will be renamed to `is_some_and()`).

Anyway, the implementation I pushed now at least gets rid of the duplicated `errors.push(...)` part, which makes me happy :)  Let me know what do you think.

Reply:
I simplified it a bit more now by replacing the trivial `if let Some(v) = val { v } else { ... }` blocks with `.unwrap_or()` and `.unwrap_or_else()` as appropriate.